feat: add functionality to search through app source tree

sudharsan-selvaraj opened the pull request:
Fixes https://github.com/appium/appium-inspector/issues/1004

**Demo**


https://github.com/appium/appium-inspector/assets/20136913/70ee6680-6504-47ac-83e3-bb1b32c709b1

eglitise commented on app/renderer/src/components/Inspector/Inspector.jsx:
```diff
@@ -314,7 +316,14 @@ const Inspector = (props) => {
+                          <Input
+                            size="middle"
+                            allowClear
+                            placeholder="Search in page source"
```
Please add this to the translations file

sudharsan-selvaraj replied:
Moved the string to translations file.

eglitise commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +21,98 @@ const Source = (props) => {
+  /**
+   * Highlights the part of the node text in source tree that matches the search term.
+   */
+  const higlightNodeMatchingSearchTerm = (nodeText, searchText) => {
```
Typo in variable name

eglitise commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +21,98 @@ const Source = (props) => {
+    const prefix = nodeText.substring(0, index);
+    const suffix = nodeText.slice(index + searchText.length);
+    //Matched word will be wrapped in a separate span for custom highlighting
+    const matchedWord = nodeText.slice(index, index + searchText.length);
```
I assume `matchedWord` is used due to potentially different casing from `searchText`?

sudharsan-selvaraj replied:
Yup. Searching is current not case sensitive. So to preserve the highlighted text casing we are using the `matchedWord` to extract the exact word from the source.

mykola-mokhnach replied:
can we just do `matchedWord = suffix`?

eglitise reviewed (changes requested):
Thank you! Will test this a bit later.
Similarly to your other PR, would it be possible to also update the documentation? I am referring to this page: https://appium.github.io/appium-inspector/latest/session-inspector/source/
* Update the top two images
* Add a new section in Application Source (between Refreshing the Source and Toggle Attributes Button)

sudharsan-selvaraj commented:
> Thank you! Will test this a bit later. Similarly to your other PR, would it be possible to also update the documentation? I am referring to this page: https://appium.github.io/appium-inspector/latest/session-inspector/source/
> 
> * Update the top two images
> * Add a new section in Application Source (between Refreshing the Source and Toggle Attributes Button)


Thanks for pointing out to the docs. I have updated all relevant images and added a new section with the gist of search functionality. Do let me know for any improvements.

mykola-mokhnach commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +21,98 @@ const Source = (props) => {
+   */
+  const highlightNodeMatchingSearchTerm = (nodeText, searchText) => {
+    if (!searchText) {
+      return nodeText;
```
It is not very clear to me why we return the nodeText if no matches have been found. I would expect nothing to be returned/highlighted in such case

mykola-mokhnach commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +21,98 @@ const Source = (props) => {
+      return nodeText;
+    }
+
+    const index = nodeText.toLowerCase().indexOf(searchText.toLowerCase());
```
is nodeText always defined?

sudharsan-selvaraj replied:
Node Text represents each XML node in the app source. So it's unlikely to be undefined. But i have added a safety check to make sure it is defined before processing.

mykola-mokhnach commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +21,98 @@ const Source = (props) => {
+       * parents to the 'nodesMatchingSearchTerm' array to make them automatically expand.
+       */
+      const nodeText = renderToString(node.title).toLowerCase();
+      if (nodeText.indexOf(pageSourceSearchText.toLowerCase()) > -1) {
```
consider using `.includes`

sudharsan-selvaraj replied:
Updated the logic.

mykola-mokhnach commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +21,98 @@ const Source = (props) => {
   } = props;
 
+  /**
+   * Highlights the part of the node text in source tree that matches the search term.
```
Please add a description for the returned value

mykola-mokhnach commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +21,98 @@ const Source = (props) => {
+  /**
+   * Highlights the part of the node text in source tree that matches the search term.
+   */
+  const highlightNodeMatchingSearchTerm = (nodeText, searchText) => {
```
it would be nice to add a unit test for this method

sudharsan-selvaraj replied:
Added necessaty unit tests.

eglitise commented on package.json:
```diff
@@ -150,6 +150,7 @@
     "eslint-plugin-promise": "6.2.0",
     "eslint-plugin-react": "7.34.2",
     "eslint-plugin-react-native": "4.1.0",
+    "ignore-styles": "^5.0.1",
```
Is this import necessary? The repo has been archived as of 2 years ago.

eglitise replied:
Hmm, I can suggest a slightly different approach: instead of exporting the whole `highlightNodeMatchingSearchTerm` function, separate its text search functionality into its own method and place it in `utils/source-parsing.js`. Something like this:
```js
export function findNodeMatchingSearchTerm(nodeText, searchText) {
  if (!searchText || !nodeText) {
    return null;
  }

  const index = nodeText.toLowerCase().indexOf(searchText.toLowerCase());
  if (index < 0) {
    return null;
  }
  const prefix = nodeText.substring(0, index);
  const suffix = nodeText.slice(index + searchText.length);
  // Matched word will be wrapped in a separate span for custom highlighting
  const matchedWord = nodeText.slice(index, index + searchText.length);
  return {prefix, matchedWord, suffix};
}
```

Then `highlightNodeMatchingSearchTerm` can be as follows:
```js
const highlightNodeMatchingSearchTerm = (nodeText, searchText) => {
  const searchResults = findNodeMatchingSearchTerm(nodeText, searchText);
  if (!searchResults) {
    return nodeText;
  }
  const {prefix, matchedWord, suffix} = searchResults;
  return (
    <>
      {prefix}
      <span className="search-word-highlighted" data-match={searchText}>
        {matchedWord}
      </span>
      {suffix}
    </>
  );
};
```

Then you can change the unit tests to validate `findNodeMatchingSearchTerm` (in `utils-source-parsing-specs.js`), and this extra import will not be necessary.

sudharsan-selvaraj replied:
Got it. i was banging my head around looking for ways to enable the testability. This make more sense and will make the necessary changes.

eglitise commented on app/renderer/src/assets/stylesheets/common.global.less:
```diff
@@ -99,3 +99,8 @@ body::-webkit-scrollbar-corner {
   font-size: 14px;
 }
+
+.search-word-highlighted {
```
I would suggest moving this back to the Inspector-specific stylesheet, since this will not be used for the Session page

sudharsan-selvaraj replied:
I moved it to common module to improve the testability and since it is a common class which can be used across any module in future it makes sense to keep it in common css.

eglitise replied:
See my other comment - if nothing is exported from the JSX, this is not needed for test purposes. Also, this file does not define any custom styles, but only overrides default ones. All custom styles are contained in the page-specific stylesheets.

sudharsan-selvaraj replied:
Got it. i was banging my head around looking for ways to enable the testability. This make more sense and will make the necessary changes.

jlipps reviewed (commented):
cool feature!

eglitise commented on app/renderer/src/components/Inspector/Source.jsx:
```diff
@@ -20,27 +52,70 @@ const Source = (props) => {
+    };
+    treeData.forEach((node) => findNodesToExpand(node, [node.key]));
+    setExpandedPaths(uniq(nodesMatchingSearchTerm));
+  }, [treeData, pageSourceSearchText]);
```
I wanted to test this but unfortunately I got an error: `Cannot access 'treeData' before initialization`.
If I moved the `useEffect` hook below the `treeData` definition, then I got another error: `Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.`
I checked and this happens due to too many updates in `treeData`.

sudharsan-selvaraj replied:
Got it. It appears that `treeData` isn't necessary as a dependency for the hook. Instead, we just need to listen for changes in the search input and re-render the tree as needed.

Interestingly, this issue didn't come up when the PR was initially created

eglitise commented on app/renderer/src/components/Inspector/Inspector.module.css:
```diff
@@ -73,6 +73,10 @@
   height: 100%;
 }
 
+.inspector-main .interaction-tab-container :global(.ant-input-affix-wrapper) {
```
This does not look good at the minimum window width:
![image](https://github.com/appium/appium-inspector/assets/37242620/6cc48f74-23f8-459a-a977-de40cc4718d8)
Although I'm not sure how to best scale this. Ideally it should shrink without affecting the card title, but the actual behavior is the opposite.

eglitise replied:
Also I would suggest defining a class for the searchbar if you intend to change its CSS. This particular line is rather long and it's not obvious what it's meant to affect.

sudharsan-selvaraj replied:
I just tweaked a bit to make it better for smaller screens. Do let me know if it looks good.

<img width="320" alt="Screenshot 2024-08-19 at 5 38 49 PM" src="https://github.com/user-attachments/assets/27420cba-16af-45d5-899b-c11f53273aa3">

eglitise replied:
It is better in that there's no linebreak, but ideally the search bar and icons should not hide the 'App Source' title - instead the title should hide the search bar/buttons. Not sure how possible this is.

eglitise reviewed (changes requested):
It seems to work well in the browser, but not so much in the Electron version :(

mykola-mokhnach commented:
@sudharsan-selvaraj Would you have some time to finish this PR?

sudharsan-selvaraj commented:
> @sudharsan-selvaraj Would you have some time to finish this PR?

@mykola-mokhnach Yup, I will be able to close the pending items.

eglitise commented:
The good thing is that it does work in the Electron version now :) But I have some issues with the tree node text extraction:
* It uses `renderToString` which is imported from `react-dom/server`, and [as per React docs](https://react.dev/reference/react-dom/server/renderToString#removing-rendertostring-from-the-client-code), it 'unnecessarily increases your bundle size and should be avoided'. There do seem to be alternatives but I haven't checked them in detail.
* The converted node titles have additional wrapper text, which is unfortunately also taken into account when searching. For example, a node `<android.widget.LinearLayout>` would be converted to `<span>&lt;<b class="_sourcetag_zwtnd_433">android.widget.linearlayout</b>&gt;</span>`. Meaning, if I search for `span`, the entire DOM tree will be unwrapped, and potentially not a single node will be highlighted.

eglitise commented:
I would also suggest exploring the idea of not converting the entire DOM upon every change in the search string. Maybe the raw title of each node can be calculated as part of `recursive`, or at least the DOM should be converted only once upon starting search.

saikrishna321 commented:
@sudharsan-selvaraj any updates on this PR ?

eglitise commented:
@sudharsan-selvaraj I will close this PR as it has been more than a year since its opening, but you are welcome to reopen it after resolving the merge conflicts and addressing the comments.
